app/src/org/commcare/android/util/CommCareExceptionHandler.java

```diff
@@ -1,26 +1,56 @@
 package org.commcare.android.util;
 
+import android.content.Context;
+import android.content.Intent;
+
 import org.commcare.android.tasks.ExceptionReportTask;
+import org.commcare.dalvik.activities.CrashWarningActivity;
+import org.javarosa.core.services.locale.Localization;
+import org.javarosa.core.util.NoLocalizedTextException;
 
 import java.lang.Thread.UncaughtExceptionHandler;
 
 /**
- * TODO: This class is basically just for testing, it should be rewritten for 
- * completeness and usefulness.
- * 
+ * Report unrecoverable exception to servers and shows user crash message
+ * for certain exceptions.  If the user is shown the crash message, they
+ * are not given the option to report a problem.
+ *
  * @author ctsims
  */
 public class CommCareExceptionHandler implements UncaughtExceptionHandler {
+    private final UncaughtExceptionHandler parent;
+    private final Context ctx;
+    public static final String WARNING_MESSAGE_KEY = "warning-message";
+    public static final String HOW_TO_FIX_MESSAGE_KEY = "fix-message";
 
-    UncaughtExceptionHandler parent;
-    
-    public CommCareExceptionHandler(UncaughtExceptionHandler parent) {
+    public CommCareExceptionHandler(UncaughtExceptionHandler parent,
+                                    Context ctx) {
         this.parent = parent;
+        this.ctx = ctx.getApplicationContext();
     }
+
     @Override
     public void uncaughtException(Thread thread, Throwable ex) {
         ExceptionReportTask task = new ExceptionReportTask();
         task.execute(ex);
-        parent.uncaughtException(thread, ex);
+
+        if (warnUserBeforeExit(ex)) {
+            System.exit(0);
+        } else {
+            parent.uncaughtException(thread, ex);
+        }
```

Comment:
Note: if the user is shown a warning they will not be given the option to report a problem. Do you agree with this design decision?